### PR TITLE
Convert pytorch models back to tensorflow

### PR DESCRIPTION
**Pull request by chrislarson1**
Added a file that converts pytorch models that have been trained/finetuned back to tensorflow. This currently supports the base BERT models (uncased/cased); conversion for other BERT models will be added in the future.

### CODE REVIEW

File: pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        """todo: compile as regex"""
+        name = name.replace('layer.',                       'layer_')
+        name = name.replace('word_embeddings.weight',       'word_embeddings')
+        name = name.replace('position_embeddings.weight',   'position_embeddings')
+        name = name.replace('token_type_embeddings.weight', 'token_type_embeddings')
+        name = name.replace('.',                            '/')
+        name = name.replace('LayerNorm/weight',             'LayerNorm/gamma')
+        name = name.replace('LayerNorm/bias',               'LayerNorm/beta')
+        name = name.replace('weight',                       'kernel')
+        # name += ':0'
```
Comment (MottoX) [inline pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py]: should remove this line

File: pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    saver.save(session, os.path.join(ckpt_dir, args.pytorch_model_name))
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--pytorch_model_dir",
+                        default=None,
+                        type=str,
+                        required=True,
```
Comment (MottoX) [inline pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py]: This is used as `cache_dir`. Why is it a required argument?

File: pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py (hunk 1 above)
Comment (chrislarson1) [reply pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py]: removed!

File: pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py (hunk 2 above)
Comment (chrislarson1) [reply pytorch_pretrained_bert/convert_hf_checkpoint_to_tf.py]: No reason. Changed now.

**Comment by chrislarson1** [comment]
Changed filename from convert_hf_checkpoint_to_tf.py to convert_pytorch_checkpoint_to_tf.py for consistency.

**Comment by Qiuzhuang** [comment]
I use this to convert the fine-tuned pytorch model to TF and convert this converted TF back to pytorch model. The prediction results seems incorrect with the converted-converted pytorch model.

**Comment by Qiuzhuang** [comment]
I compare the dumped stat_dict, it seems the differences lie in :

encoder.layer.{d}.attention.self.query.weight
encoder.layer.{d}.attention.self.key.weight
encoder.layer.{d}.attention.self.value.weight

**Comment by chrislarson1** [comment]
What model are you trying to convert @Qiuzhuang? Per the docstring, only the BertModel is currently supported. I will add support other models in the near future.

**Comment by Qiuzhuang** [comment]
Hi @chrislarson1, I use BertModel as follows:
BertModel(config).from_pretrained(pretrain_model_dir)
where pretrain_model_dir is (domain-pretraining + task specific classifier) training model dir.

**Comment by Qiuzhuang** [comment]
I write the sample test code read the pytorch model and tf-pytorch model via BertModel.

**Comment by Qiuzhuang** [comment]
@chrislarson1 we need to transpose attention Q/K/V weight as well,  here is the fixing:

        if any(pattern in var_name for pattern in ["dense.weight", "attention.self.query", "attention.self.key", "attention.self.value"]):
            torch_tensor = torch_tensor.T
        tf_tensor = assign_tf_var(tensor=torch_tensor, name=tf_name)

**Comment by Qiuzhuang** [comment]
I am able to train pytorch fine-tuned model and then convert to tensorflow model for serving purpose. E.g. using bert-as-service. The results are very consistent.

**Comment by chrislarson1** [comment]
Thanks @Qiuzhuang, your change has been added.

**Comment by thomwolf** [comment]
Thanks, @chrislarson1 and @Qiuzhuang.
This feature in interesting indeed.
I think it could be nice to add a check that output of the PyTorch and TensorFlow models are identical (on at least one example). For the Bert model, I did a simple notebook (see the notebook folder) but it can also be a script or a test.
Do you think you can add something like that @chrislarson1 ?

**Comment by chrislarson1** [comment]
No problem @thomwolf, I've added a notebook that checks correctness.

**Comment by thomwolf** [comment]
Ok great, let's merge it!

**Comment by justachetan** [comment]
Hi, why did we take a transpose of the layers here at all? Is it because of tensorflow treating its layers differently than pytorch?

Also, if the dense.weight layers are being transposed, then in s sequence classification model, will the out_proj layer need to be transposed as well?

**Comment by chrislarson1** [comment]
Hi @justachetan, the answer lies in how linear transformations are represented in pytorch vs. tensorflow; they are not the same. In Pytorch, weights in a network **often** get wrapped in the torch.nn.Linear class, which store transposed versions of the weights that would get saved in tensorflow for an equivalent projection (see the example below).

```
>>> import numpy as np
>>> import tensorflow as tf
>>> import torch
>>> from torch.functional import F
>>> x = np.ones([10, 4])
>>> W = np.ones([5, 4])
>>> tf.matmul(x, W.T).shape
(10, 5)
>>> F.linear(torch.Tensor(x), torch.Tensor(W)).detach().numpy().shape
(10, 5)
```

**Pull request merged**